File: tests/cross_driver/java/src/main/java/com/secantus/smokes/BulkSmoke.java
package com.secantus.smokes;

import com.mongodb.ConnectionString;
import com.mongodb.MongoClientSettings;
import com.mongodb.bulk.BulkWriteResult;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.model.DeleteOneModel;
import com.mongodb.client.model.InsertOneModel;
import com.mongodb.client.model.ReplaceOneModel;
import com.mongodb.client.model.ReplaceOptions;
import com.mongodb.client.model.UpdateManyModel;
import com.mongodb.client.model.UpdateOneModel;
import com.mongodb.client.model.UpdateOptions;
import com.mongodb.client.model.WriteModel;
import org.bson.Document;
import org.bson.conversions.Bson;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Cross-driver bulk-write smoke — Java.
 * <p>
 * One mixed bulkWrite that insert / update / replace / upsert /
 * delete-walks across six docs, then asserts the result counts and
 * final collection state. The Java driver's bulk command builder is
 * a separate implementation from pymongo's.
 */
public final class BulkSmoke {

    private BulkSmoke() {}

    public static void main(String[] args) {
        String uri = System.getenv("MONGODB_URI");
        if (uri == null) {
            System.err.println("MONGODB_URI not set");
            System.exit(2);
        }

        MongoClientSettings settings = MongoClientSettings.builder()
            .applyConnectionString(new ConnectionString(uri))
            .applyToClusterSettings(b -> b.serverSelectionTimeout(30, TimeUnit.SECONDS))
            .build();

        try (MongoClient client = MongoClients.create(settings)) {
            MongoCollection<Document> coll = client.getDatabase("bulk_xd").getCollection("c");
            coll.drop();

            coll.insertMany(List.of(
                new Document("_id", 1).append("kind", "old"),
                new Document("_id", 2).append("kind", "old")));

            List<WriteModel<Document>> ops = new ArrayList<>();
            ops.add(new InsertOneModel<>(new Document("_id", 3).append("kind", "fresh")));
            ops.add(new UpdateOneModel<>(
                new Document("_id", 1),
                new Document("$set", new Document("kind", "new"))));
            ops.add(new UpdateManyModel<>(
                new Document("kind", "old"),
                new Document("$set", new Document("kind", "new"))));
            ops.add(new ReplaceOneModel<>(
                new Document("_id", 3),
                new Document("_id", 3).append("kind", "replaced"),
                new ReplaceOptions()));
            ops.add(new UpdateOneModel<>(
                new Document("_id", 99),
                new Document("$set", new Document("kind", "upserted")),
                new UpdateOptions().upsert(true)));
            ops.add(new DeleteOneModel<>(new Document("_id", 2)));

            BulkWriteResult res = coll.bulkWrite(ops);

            if (res.getInsertedCount() != 1) {
                System.err.printf("insertedCount: got %d, want 1%n", res.getInsertedCount());
                System.exit(1);
            }
            if (res.getMatchedCount() != 3) {
                System.err.printf("matchedCount: got %d, want 3%n", res.getMatchedCount());
                System.exit(1);
            }
            if (res.getModifiedCount() != 3) {
                System.err.printf("modifiedCount: got %d, want 3%n", res.getModifiedCount());
                System.exit(1);
            }
            if (res.getUpserts().size() != 1) {
                System.err.printf("upsertedCount: got %d, want 1%n", res.getUpserts().size());
                System.exit(1);
            }
            if (res.getDeletedCount() != 1) {
                System.err.printf("deletedCount: got %d, want 1%n", res.getDeletedCount());
                System.exit(1);
            }

            List<Document> got = new ArrayList<>();
            Bson sort = new Document("_id", 1);
            coll.find(new Document()).sort(sort).into(got);
            if (got.size() != 3) {
                System.err.printf("final docs: got %d, want 3%n", got.size());
                System.exit(1);
            }
            if (got.get(0).getInteger("_id") != 1 || !"new".equals(got.get(0).getString("kind"))) {
                System.err.printf("doc[0]: got %s%n", got.get(0).toJson());
                System.exit(1);
            }
            if (got.get(1).getInteger("_id") != 3 || !"replaced".equals(got.get(1).getString("kind"))) {
                System.err.printf("doc[1]: got %s%n", got.get(1).toJson());
                System.exit(1);
            }
            if (got.get(2).getInteger("_id") != 99 || !"upserted".equals(got.get(2).getString("kind"))) {
                System.err.printf("doc[2]: got %s%n", got.get(2).toJson());
                System.exit(1);
            }
            System.out.println("OK");
        }
    }
}

File: tests/cross_driver/php/bulk_smoke.php
<?php
// Cross-driver bulk-write smoke — PHP (mongo-php-library).
//
// One mixed bulkWrite that insert / update / replace / upsert /
// delete-walks across six docs, then asserts the result counts and
// final collection state.

declare(strict_types=1);

require __DIR__ . '/vendor/autoload.php';

use MongoDB\Client;

function bail(string $msg): never {
    fwrite(STDERR, $msg . PHP_EOL);
    exit(1);
}

$uri = getenv('MONGODB_URI') ?: bail('MONGODB_URI not set');

$client = new Client($uri, ['serverSelectionTimeoutMS' => 5000]);
$coll = $client->bulk_xd_php->c;
$coll->drop();

$coll->insertMany([
    ['_id' => 1, 'kind' => 'old'],
    ['_id' => 2, 'kind' => 'old'],
]);

$res = $coll->bulkWrite([
    ['insertOne' => [['_id' => 3, 'kind' => 'fresh']]],
    ['updateOne' => [['_id' => 1], ['$set' => ['kind' => 'new']]]],
    ['updateMany' => [['kind' => 'old'], ['$set' => ['kind' => 'new']]]],
    ['replaceOne' => [['_id' => 3], ['_id' => 3, 'kind' => 'replaced']]],
    ['updateOne' => [['_id' => 99], ['$set' => ['kind' => 'upserted']], ['upsert' => true]]],
    ['deleteOne' => [['_id' => 2]]],
]);

if ($res->getInsertedCount() !== 1) bail('insertedCount: got ' . $res->getInsertedCount() . ', want 1');
if ($res->getMatchedCount() !== 3) bail('matchedCount: got ' . $res->getMatchedCount() . ', want 3');
if ($res->getModifiedCount() !== 3) bail('modifiedCount: got ' . $res->getModifiedCount() . ', want 3');
if ($res->getUpsertedCount() !== 1) bail('upsertedCount: got ' . $res->getUpsertedCount() . ', want 1');
if ($res->getDeletedCount() !== 1) bail('deletedCount: got ' . $res->getDeletedCount() . ', want 1');

$got = iterator_to_array($coll->find([], ['sort' => ['_id' => 1]]));
if (count($got) !== 3) bail('final docs: got ' . count($got) . ', want 3');
$g0 = (array) $got[0];
$g1 = (array) $got[1];
$g2 = (array) $got[2];
if ((int)$g0['_id'] !== 1 || $g0['kind'] !== 'new') bail('doc[0]: ' . json_encode($g0));
if ((int)$g1['_id'] !== 3 || $g1['kind'] !== 'replaced') bail('doc[1]: ' . json_encode($g1));
if ((int)$g2['_id'] !== 99 || $g2['kind'] !== 'upserted') bail('doc[2]: ' . json_encode($g2));

echo "OK" . PHP_EOL;
